From fca1989f78a93d742956ea19c91025e8edc623b7 Mon Sep 17 00:00:00 2001
From: Mao Minkai <maominkai@wxiat.com>
Date: Fri, 12 Aug 2022 15:24:06 +0800
Subject: [PATCH] sw64: optimize instruction usage in fork routine
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: fca1989f78a93d742956ea19c91025e8edc623b7
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I5PNDF

--------------------------------

Use 'call' instead of 'jmp' in ret_from_fork(), so the 'ret' in
schedule_tail() won't mess up branch prediction.

Signed-off-by: Mao Minkai <maominkai@wxiat.com>
Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/kernel/entry.S | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/sw_64/kernel/entry.S b/arch/sw_64/kernel/entry.S
index a52665c3bb08..67bafd4a930a 100644
--- a/arch/sw_64/kernel/entry.S
+++ b/arch/sw_64/kernel/entry.S
@@ -426,7 +426,7 @@ __switch_to:
 	.ent ret_from_fork
 ret_from_fork:
 	ldi	$26, ret_from_sys_call
-	jmp	$31, schedule_tail
+	call	$31, schedule_tail
 	.end ret_from_fork
 
 /*
-- 
2.33.0

